Enabling code formatting fails in nested projects

CMake configuration fails if formatting is enabled in the case that `hpc-coding-conventions` is used at multiple "depths" in the project structure. For example NEURON has the structure
```
nrn/external/coreneuron/
                        CMake/hpc-coding-conventions
                        external/nmodl/cmake/hpc-coding-conventions
```
and trying to enable code formatting in CoreNEURON gives:
```
CMake Error at external/coreneuron/external/nmodl/cmake/hpc-coding-conventions/cpp/CMakeLists.txt:122 (add_custom_target):
  add_custom_target cannot create target "clang-format" because another
  target with the same name already exists.  The existing target is a custom
  target created in source directory
  "/gpfs/bbp.cscs.ch/home/olupton/nrn/external/coreneuron/CMake/hpc-coding-conventions/cpp".
  See documentation for policy CMP0002 for more details.
Call Stack (most recent call first):
  external/coreneuron/external/nmodl/cmake/hpc-coding-conventions/cpp/CMakeLists.txt:291 (cpp_cc_enable_clang_formatting)
```
using
```
-DNRN_ENABLE_CORENEURON=ON -DCORENRN_ENABLE_NMODL=ON -DCORENRN_CLANG_FORMAT=ON
```

I have had to use a separate, standalone, checkout of CoreNEURON to successfully run `cmake --build . --target clang-format`.

We make extensive use of submodules in our projects, and include `hpc-coding-conventions` in many projects, so I think it's important that this is usable.

(Note that (I suspect) the problem here is with the prefix handling, so `CORENRN_CLANG_FORMAT` might be trying to enable formatting in NMODL too.)

--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -63,6 +63,15 @@ function(cpp_cc_setup_tool_config name path)
   endif()
 endfunction()
 
+function(cpp_cc_add_tool_target name)
+  add_custom_target(${name}-${PROJECT_NAME} ${ARGN})
+  if(TARGET ${name})
+    add_dependencies(${name} ${name}-${PROJECT_NAME})
+  else()
+    add_custom_target(${name} DEPENDS ${name}-${PROJECT_NAME})
+  endif()
+endfunction()
+
 function(cpp_cc_enable_clang_formatting)
   find_package(PythonInterp 3.5 REQUIRED)
   set(ClangFormat_MIN_VERSION 7)
@@ -97,17 +106,19 @@ function(cpp_cc_enable_clang_formatting)
       "${PROJECT_SOURCE_DIR}" "--executable=${ClangFormat_EXECUTABLE}" --files-re
       ${${CODING_CONV_PREFIX}_ClangFormat_FILES_RE} --excludes-re
       ${${CODING_CONV_PREFIX}_ClangFormat_EXCLUDES_RE} --)
-  add_custom_target(clang-format ${clang_format_command_prefix} format --
-                                 "${${CODING_CONV_PREFIX}_ClangFormat_OPTIONS}")
-  add_custom_target(check-clang-format ${clang_format_command_prefix} check --
-                                       "${${CODING_CONV_PREFIX}_ClangFormat_OPTIONS}")
+  cpp_cc_add_tool_target(clang-format ${clang_format_command_prefix} format --
+                         "${${CODING_CONV_PREFIX}_ClangFormat_OPTIONS}")
+  cpp_cc_add_tool_target(check-clang-format ${clang_format_command_prefix} check --
+                         "${${CODING_CONV_PREFIX}_ClangFormat_OPTIONS}")
   if(${CODING_CONV_PREFIX}_TEST_FORMATTING)
-    add_test(NAME ClangFormat COMMAND ${clang_format_command_prefix} --make-unescape-re check --
-                                      "${${CODING_CONV_PREFIX}_ClangFormat_OPTIONS}")
+    add_test(NAME ClangFormat_${PROJECT_NAME}
+             COMMAND ${clang_format_command_prefix} --make-unescape-re check --
+                     "${${CODING_CONV_PREFIX}_ClangFormat_OPTIONS}")
   endif()
   if(${CODING_CONV_PREFIX}_ClangFormat_DEPENDENCIES)
-    add_dependencies(clang-format ${${CODING_CONV_PREFIX}_ClangFormat_DEPENDENCIES})
-    add_dependencies(check-clang-format ${${CODING_CONV_PREFIX}_ClangFormat_DEPENDENCIES})
+    add_dependencies(clang-format-${PROJECT_NAME} ${${CODING_CONV_PREFIX}_ClangFormat_DEPENDENCIES})
+    add_dependencies(check-clang-format-${PROJECT_NAME}
+                     ${${CODING_CONV_PREFIX}_ClangFormat_DEPENDENCIES})
   endif()
 endfunction(cpp_cc_enable_clang_formatting)
 
@@ -141,13 +152,14 @@ function(cpp_cc_enable_cmake_formatting)
       "${PROJECT_SOURCE_DIR}" "--executable=${CMakeFormat_EXECUTABLE}" --files-re
       ${${CODING_CONV_PREFIX}_CMakeFormat_FILES_RE} --excludes-re
       ${${CODING_CONV_PREFIX}_CMakeFormat_EXCLUDES_RE} --)
-  add_custom_target(cmake-format ${cmake_format_command_prefix} format --
-                                 "${${CODING_CONV_PREFIX}_CMakeFormat_OPTIONS}")
-  add_custom_target(check-cmake-format ${cmake_format_command_prefix} check --
-                                       ${${CODING_CONV_PREFIX}_CMakeFormat_OPTIONS})
+  cpp_cc_add_tool_target(cmake-format ${cmake_format_command_prefix} format --
+                         "${${CODING_CONV_PREFIX}_CMakeFormat_OPTIONS}")
+  cpp_cc_add_tool_target(check-cmake-format ${cmake_format_command_prefix} check --
+                         ${${CODING_CONV_PREFIX}_CMakeFormat_OPTIONS})
   if(${CODING_CONV_PREFIX}_TEST_FORMATTING)
-    add_test(NAME CMakeFormat COMMAND ${cmake_format_command_prefix} --make-unescape-re check --
-                                      ${${CODING_CONV_PREFIX}_CMakeFormat_OPTIONS})
+    add_test(NAME CMakeFormat_${PROJECT_NAME}
+             COMMAND ${cmake_format_command_prefix} --make-unescape-re check --
+                     ${${CODING_CONV_PREFIX}_CMakeFormat_OPTIONS})
   endif()
 
 endfunction(cpp_cc_enable_cmake_formatting)
@@ -182,14 +194,14 @@ function(cpp_cc_enable_static_analysis)
       "${${CODING_CONV_PREFIX}_ClangTidy_FILES_RE}" --excludes-re
       ${${CODING_CONV_PREFIX}_ClangTidy_EXCLUDES_RE} -p ${PROJECT_BINARY_DIR}/compile_commands.json
       check)
-  add_custom_target(clang-tidy ${clang_tidy_command_prefix} --
-                               ${${CODING_CONV_PREFIX}_ClangTidy_OPTIONS})
+  cpp_cc_add_tool_target(clang-tidy ${clang_tidy_command_prefix} --
+                         ${${CODING_CONV_PREFIX}_ClangTidy_OPTIONS})
   if(${CODING_CONV_PREFIX}_TEST_STATIC_ANALYSIS)
-    add_test(NAME ClangTidy COMMAND ${clang_tidy_command_prefix} --make-unescape-re --
-                                    ${${CODING_CONV_PREFIX}_ClangTidy_OPTIONS})
+    add_test(NAME ClangTidy_${PROJECT_NAME} COMMAND ${clang_tidy_command_prefix} --make-unescape-re
+                                                    -- ${${CODING_CONV_PREFIX}_ClangTidy_OPTIONS})
   endif()
   if(${CODING_CONV_PREFIX}_ClangTidy_DEPENDENCIES)
-    add_dependencies(clang-tidy ${${CODING_CONV_PREFIX}_ClangTidy_DEPENDENCIES})
+    add_dependencies(clang-tidy_${PROJECT_NAME} ${${CODING_CONV_PREFIX}_ClangTidy_DEPENDENCIES})
   endif()
 
 endfunction(cpp_cc_enable_static_analysis)

